torch/csrc/onnx/export.cpp

```diff
@@ -256,18 +256,46 @@ static void addAttribute(onnx::NodeProto * n_p, jit::Node * n, jit::Symbol name)
       break;
     case AttributeKind::g: {
       auto g = attr->mutable_g();
-      encodeGraph(g, n->g(name), {});
+      encodeGraph(g, n->g(name), {}, {}, {});
     } break;
     case AttributeKind::gs:
       for(auto & v : n->gs(name)) {
         auto g = attr->add_graphs();
-        encodeGraph(g, v, {});
+        encodeGraph(g, v, {}, {}, {});
       }
       break;
   }
 }
 
-static void encodeGraph(onnx::GraphProto * p_g, std::shared_ptr<Graph> & g, const std::vector<at::Tensor> & initializers) {
+
+static void encodeGraph(onnx::GraphProto * p_g, std::shared_ptr<Graph> & g,
+  const std::vector<std::string> & input_names,
+  const std::vector<std::string> & output_names,
+  const std::vector<at::Tensor> & initializers) {
+
+  std::unordered_map<Node *, std::string> node_names;
+  auto initNames = [&](const node_list & nodes, const std::vector<std::string> & names) {
+    JIT_ASSERT(names.size() <= nodes.size());
+    size_t idx = nodes.size() - names.size();
+    for (auto & name : names) {
+      auto node = nodes.at(idx++);
+      auto it = node_names.find(node);
+      // rarely, output nodes may be repeated
+      // if this happens, we just take the first name we find for it as its name
+      if(it == node_names.end()) {
+        node_names[node] = name;
```

Comment:
If this case triggers, the export process will silently fail to give a correct name to the output. When a user attempts to run the model from ONNX, they will simply be told that a name (they rightly expect to exist) doesn't exist. You won't be able to debug this situation without looking at the protobuf.

It would be much better to just raise an error in this case.

Reply:
This will make it impossible to use output_names on any network that does this. Is that OK?

Reply:
Well, as I mentioned during our check-in today, if we were going to fix this "properly", we'd open a PR to ONNX to add named outputs as an extra piece of metadata in the output. Then we could rightfully only be in the business of assigning to 'env' at binding sites (input is a binding site; output is NOT) and both the "intermediate result returned twice" and the "input returned directly as output" cases would work correctly. But this involves doing the onnx runaround...

Absent that, I think erroring in this case is pretty darn reasonable. I mean, it's not like output_names is going to "work" anyway, if the user actually wants to use the named outputs on the other end.